pkg/controllers/ttl-controller/utils.go

```diff
@@ -0,0 +1,70 @@
+package ttlcontroller
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	checker "github.com/kyverno/kyverno/pkg/auth/checker"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+)
+
+func discoverResources(logger logr.Logger, discoveryClient discovery.DiscoveryInterface) ([]schema.GroupVersionResource, error) {
+	var resources []schema.GroupVersionResource
+	apiResourceList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		if !discovery.IsGroupDiscoveryFailedError(err) {
+			return nil, err
+		}
+		// the error should be recoverable, let's log missing groups and process the partial results we received
+		err := err.(*discovery.ErrGroupDiscoveryFailed)
+		for gv, err := range err.Groups {
+			// Handling the specific group error
+			logger.Error(err, "error in discovering group", "gv", gv)
+		}
+	}
+	for _, apiResourceList := range apiResourceList {
+		for _, apiResource := range apiResourceList.APIResources {
+			if sets.NewString(apiResource.Verbs...).HasAll("list", "watch", "delete") {
+				groupVersion, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+				if err != nil {
+					return resources, err
+				}
+				resources = append(resources, groupVersion.WithResource(apiResource.Name))
+			}
+		}
+	}
+	return resources, nil
+}
+
+func hasResourcePermissions(logger logr.Logger, resource schema.GroupVersionResource, s checker.AuthChecker) bool {
+	can, err := checker.Check(context.TODO(), s, resource.Group, resource.Version, resource.Resource, "", "", "watch", "list", "delete")
+	if err != nil {
+		logger.Error(err, "failed to check permissions")
+		return false
+	}
+	return can
+}
+
+func parseDeletionTime(metaObj metav1.Object, deletionTime *time.Time, ttlValue string) error {
+	ttlDuration, err := time.ParseDuration(ttlValue)
+	if err == nil {
+		creationTime := metaObj.GetCreationTimestamp().Time
+		*deletionTime = creationTime.Add(ttlDuration)
+	} else {
+		layoutRFCC := "2006-01-02T150405Z"
+		// Try parsing ttlValue as a time in ISO 8601 format
+		*deletionTime, err = time.Parse(layoutRFCC, ttlValue)
```

Comment:
The valid format of the timestamp needs to be documented properly.

Reply:
I will create the corresponding issue for docs today.

Reply:
Done